api/logic/evm_test.go

```diff
@@ -0,0 +1,93 @@
+package logic
+
+import (
+	_ "embed"
+	"encoding/hex"
+	"encoding/json"
+	"math/big"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	ethCommon "github.com/ethereum/go-ethereum/common"
+	"github.com/stretchr/testify/require"
+
+	"github.com/oasisprotocol/oasis-indexer/analyzer/evmabi"
+)
+
+//go:embed test_contracts/artifacts/Varied.json
+var artifactVariedJSON []byte
+var Varied *abi.ABI
+
+func init() {
+	type artifact struct {
+		ABI *abi.ABI
+	}
+	var artifactVaried artifact
+	if err := json.Unmarshal(artifactVariedJSON, &artifactVaried); err != nil {
+		panic(err)
+	}
+	Varied = artifactVaried.ABI
+}
+
+func TestEVMParseTypes(t *testing.T) {
+	data, err := hex.DecodeString("a85acffaffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000000000000000000000000000000000000000000001ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000010101010101010101010101010101010101010101010101010101010101010101000000000000000000000000010101010101010101010101010101010101010101010101010101010101010101010101010101010202020200000000000000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000001c00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000024000000000000000000000000000000000000000000000000000000000000002800000000000000000000000000000000000000000000000000000000000000001010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000016100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000016100000000000000000000000000000000000000000000000000000000000000")
+	require.NoError(t, err)
+	method, args, err := EVMParseData(data, Varied)
+	require.Equal(t, Varied.Methods["test"], *method)
+	jsonExpected := []string{
+		"-1",     // int8
+		"1",      // uint8
+		"\"-1\"", // int256
+		"\"1\"",  // uint256
+		"true",   // bool
+		"\"AQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQE=\"", // bytes32
+		"\"0x0101010101010101010101010101010101010101\"",   // address
+		"\"AQEBAQEBAQEBAQEBAQEBAQEBAQECAgIC\"",             // function (uint16) external returns (uint16)
```

Comment:
Not important (because fns won't get passed/emitted much, I reckon), but out of curiosity, is this the fn ... body? address of some kind? signature of a hash that includes the name?
My guess is the latter, in which case it would be nicer to display the fn name. But if it involves indpendent lookups into the ABI, let's not go down that rabbithole and just format this in hex.

Reply:
it's contract addr and that 4-byte solidity function selector